lib/BuildSystem/BuildNode.cpp

```diff
@@ -133,7 +133,15 @@ bool BuildNode::configureAttribute(
 
 FileInfo BuildNode::getFileInfo(basic::FileSystem& fileSystem) const {
   assert(!isVirtual());
-  return fileSystem.getFileInfo(getName());
+
+  // Drop the trailing slash
+  // otherwise non-directory paths that end with "/" will be reported as missing
+  StringRef path = getName();
+  if (path.endswith("/") && path != "/") {
```

Comment:
Even with the explanatory comments, these checks are very subtle. I wonder if there's a way we could model this better (maybe there isn't)

Reply:
I'm not certain, but this might be unavoidable if the high-level build system is encoding all input/output nodes as directory in the manifest.

When high-level build system treats all nodes as directories (regardless of whether user typed in a trailing slash or not), it introduces ambiguity in user's intent, without providing any opportunity for disambiguation.

At the time `ExternalCommand::computeCommandResult` is calling `node->getFileInfo(..)` for each of its outputs, llbuild can't distinguish whether the output node was truly intended to be a file or a directory.

If we want to avoid this check, the high-level build system needs to allow user to express their intent unambiguously. (i.e. high-level build system should only encode nodes with explicit trailing slash as directory node)..

@m-i-r-z-a what do you think?